EZP-30773: PHP API: Search (#714)
ezsystems/developer-documentation: docs/api/public_php_api_search.md

You can also use [`SearchService::findContent`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/SearchService.php#L122) to get full Content objects, together with their Field information.

You can also query for a single result, for example by providing a Content ID,
with the [`SearchService::findSingle`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/SearchService.php#L159) method:

``` php
$criterion = new Criterion\ContentId($contentId);
$result = $this->searchService->findSingle($criterion);
$output->writeln($result->getName());
```

!!! tip

    For full list and details of available Search Criteria, see [Search Criteria reference](../guide/search.md#search-criteria-reference).

### Search with `query` and `filter`

You can use two properties of the `Query` object to search for Content: `query` and `filter`.

In contrast to `filter`, `query` has an effect of search scoring (relevancy).
It affects default sorting if no Sort Clause is used.

> As such, `query` is recommended when the search uses is based on user input.

Reviewer: ```suggestion
As such, `query` is recommended when the search uses are based on user input.
```

> The difference between `query` and `filter` is only in play when using Solr search engine.

Reviewer: ```suggestion
The difference between `query` and `filter` is only relevant when using Solr search engine.
```

With the Legacy search engine both properties will give identical results.

## Searching in a controller

You can use the `SearchService` similarly in a controller, as long as you provide the required parameters.
For example, in the code below `locationId` is provided to list all children of a Location.

``` php hl_lines="20 21 22"
//...
use eZ\Publish\API\Repository\SearchService;
use eZ\Publish\API\Repository\Values\Content\LocationQuery;
use eZ\Publish\API\Repository\Values\Content\Query\Criterion;

class CustomController extends Controller
{
    //...
    public function showContentAction($locationId)
    {
        $query = new LocationQuery();
        $query->filter = new Criterion\ParentLocationId($locationId);

        $results = $this->searchService->findContentInfo($query);
        $items = [];
        foreach ($results->searchHits as $searchHit) {
            $items[] = $searchHit;
        }

        return $this->render('custom.html.twig', [
            'items' => $items,
        ]);
    }
}
```

The rendering of results is then relegated to [templates](../guide/templates.md) (lines 20-22).

### Paginating search results

> To paginate search results, it is recommended to use the [Pagerfanta library](https://github.com/whiteoctober/Pagerfanta) and [eZ Platform's adapters for it](https://github.com/ezsystems/ezpublish-kernel/tree/v7.5.2/eZ/Publish/Core/Pagination/Pagerfanta).

Reviewer: ```suggestion
To paginate search results, it is recommended to use the [Pagerfanta library](https://github.com/whiteoctober/Pagerfanta) and [eZ Platform's adapters for it.](https://github.com/ezsystems/ezpublish-kernel/tree/v7.5.2/eZ/Publish/Core/Pagination/Pagerfanta)
```

``` php
//...
use eZ\Publish\Core\Pagination\Pagerfanta\ContentSearchAdapter;
use Symfony\Component\HttpFoundation\Request;
use Pagerfanta\Pagerfanta;

class CustomController extends Controller
{
    //...
    public function showContentAction(Request $request, $locationId)
    {
        // formulate a $query

        $pager = new Pagerfanta(
            new ContentSearchAdapter($query, $this->searchService)
        );
        $pager->setMaxPerPage(3);
        $pager->setCurrentPage($request->get('page', 1));

        return $this->render('custom.html.twig', [
                'totalItemCount' => $pager->getNbResults(),
                'pagerItems' => $pager,
            ]
        );
    }
}
```

Pagination can then be rendered for example using the following template:

``` html+twig
{% for item in pagerItems %}
    <h2><a href={{ path('ez_urlalias', {'contentId': item.contentInfo.id}) }}>{{ ez_content_name(item) }}</a></h2>
{% endfor %}

{% if pagerItems.haveToPaginate() %}
    {{ pagerfanta( pagerArticle, 'ez') }}
{% endif %}
```

> For more information and examples, see [PagerFanta documentation](https://github.com/whiteoctober/Pagerfanta/blob/master/README.md).

Reviewer: ```suggestion
For more information and examples, see [PagerFanta documentation.](https://github.com/whiteoctober/Pagerfanta/blob/master/README.md)
```

foreach ($result->searchHits as $searchHit) {
    $output->writeln($searchHit->valueObject->name);
}
```

This example takes three parameters from a command — `$text`, `$contentTypeId`, and `$locationId`.
It then combines them using `Criterion\LogicalAnd` to search for Content items
that belong to a specific subtree, have the chosen Content Type and contain the provided text (lines 6-8).

This also shows that you can get the total number of search results using the `totalCount` property of search results (line 11).

You can also nest different operators to construct more complex queries.
The example below uses the `LogicalNot` operator to search for all children of the selected parent
that do not belong to the provided Content Type:

``` php
$query->filter = new Criterion\LogicalAnd([
        new Criterion\ParentLocationId($locationId),
        new Criterion\LogicalNot(
            new Criterion\ContentTypeIdentifier($contentTypeId)
        )
    ]
);
```

### Combining independent Criteria

Criteria are independent of one another. This can lead to unexpected behavior, for instance because Content can have multiple Locations.

> For example, a Content item has two locations: visible Location A and hidden Location B.

Reviewer: ```suggestion
For example, a Content item has two Locations: visible Location A and hidden Location B.
```

        $output->writeln('Found ' . $result->totalCount . ' items');
        foreach ($result->searchHits as $searchHit) {
            $output->writeln($searchHit->valueObject->name);
        }
    }
}
```

[`SearchService::findContentInfo`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/SearchService.php#L142) (line 16)
retrieves [`ContentInfo`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/Values/Content/ContentInfo.php) objects of the found Content items.
You can also use [`SearchService::findContent`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/SearchService.php#L122) to get full Content objects, together with their Field information.

You can also query for a single result, for example by providing a Content ID,
with the [`SearchService::findSingle`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/SearchService.php#L159) method:

``` php
$criterion = new Criterion\ContentId($contentId);
$result = $this->searchService->findSingle($criterion);
$output->writeln($result->getName());
```

!!! tip

>     See [Search Criteria reference](../guide/search.md#search-criteria-reference) for full list and details of available Search Criteria.

Reviewer 2: ```suggestion
    For full list and details of available Search Criteria, see [Search Criteria reference](../guide/search.md#search-criteria-reference).
```

### Search with `query` and `filter`

You can use two properties of the `Query` object to search for Content: `query` and `filter`.

In contrast to `filter`, `query` has an effect of search scoring (relevancy).
It affects default sorting if no Sort Clause is used.
As such, `query` is recommended when the search uses is based on user input.

The difference between `query` and `filter` is only in play when using Solr search engine.

> With Legacy search engine both properties will give identical results.

Reviewer 2: ```suggestion
With the Legacy search engine both properties will give identical results.
```

## Searching in a controller

You can use the `SearchService` similarly in a controller, as long as you provide the required parameters

> (in the example below, `locationId` is provided to list all children of a Location).

Reviewer 2: ```suggestion
As the example below demonstrates, `locationId` is provided to list all children of a Location.
```

``` php hl_lines="20 21 22"
//...
use eZ\Publish\API\Repository\SearchService;
use eZ\Publish\API\Repository\Values\Content\LocationQuery;
use eZ\Publish\API\Repository\Values\Content\Query\Criterion;

class CustomController extends Controller
{
    //...
    public function showContentAction($locationId)
    {
        $query = new LocationQuery();
        $query->filter = new Criterion\ParentLocationId($locationId);

        $results = $this->searchService->findContentInfo($query);
        $items = [];
        foreach ($results->searchHits as $searchHit) {
            $items[] = $searchHit;
        }

        return $this->render('custom.html.twig', [
            'items' => $items,
        ]);
    }
}
```

> Rendering results is then relegated to [templates](../guide/templates.md) (lines 20-22).

Reviewer 2: ```suggestion
Rendering results are then relegated to [templates](../guide/templates.md) (lines 20-22).
```

|[`ContentSearchHitAdapter`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/Core/Pagination/Pagerfanta/ContentSearchHitAdapter.php)|Makes a search against passed Query and returns [SearchHit](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/Values/Content/Search/SearchHit.php) objects instead.|
|[`LocationSearchAdapter`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/Core/Pagination/Pagerfanta/LocationSearchAdapter.php)|Makes a Location search against passed Query and returns [Location](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/Values/Content/Location.php) objects.|
|[`LocationSearchHitAdapter`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/Core/Pagination/Pagerfanta/LocationSearchHitAdapter.php)|Makes a Location search against passed Query and  returns [SearchHit](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/Values/Content/Search/SearchHit.php) objects instead.|

## Complex search

For more complex searches, you need to combine multiple Criteria.

> You can do it using logical operators `LogicalAnd`, `LogicalOr` and `LogicalNot`.

Reviewer 2: ```suggestion
You can do it using logical operators: `LogicalAnd`, `LogicalOr`, and `LogicalNot`.
```

``` php hl_lines="6 7 8 11"
$query = new LocationQuery;
$criterion1 = new Criterion\Subtree($this->locationService->loadLocation($locationId)->pathString);
$criterion2 = new Criterion\ContentTypeId($contentTypeId);
$criterion3 = new Criterion\FullText($text);

$query->query = new Criterion\LogicalAnd(
    [$criterion1, $criterion2, $criterion3]
);

$result = $this->searchService->findContentInfo($query);
$output->writeln('Found ' . $result->totalCount . ' items');
foreach ($result->searchHits as $searchHit) {
    $output->writeln($searchHit->valueObject->name);
}
```

> This example takes three parameters from a command, `$text`, `$contentTypeId`, and `$locationId`.

Reviewer 2: ```suggestion
This example takes three parameters from a command — `$text`, `$contentTypeId`, and `$locationId`.
```

This also shows that you can get the total number of search results using the `totalCount` property of search results (line 11).

You can also nest different operators to construct more complex queries.
The example below uses the `LogicalNot` operator to search for all children of the selected parent
that do not belong to the provided Content Type:

``` php
$query->filter = new Criterion\LogicalAnd([
        new Criterion\ParentLocationId($locationId),
        new Criterion\LogicalNot(
            new Criterion\ContentTypeIdentifier($contentTypeId)
        )
    ]
);
```

### Combining independent Criteria

Criteria are independent of one another. This can lead to unexpected behavior, for instance because Content can have multiple Locations.

For example, a Content item has two locations: visible Location A and hidden Location B.
You perform the following query:

``` php
$query->filter = new Criterion\LogicalAnd([
    new LocationId($locationBId),
    new Visibility(Visibility::VISIBLE),
]);
```

The query searches for Location B using the `LocationId` Criterion,
and for visible Content using the `Visibility` Criterion.

> The query will find the Content because both conditions are satisfied:

Reviewer 2: ```suggestion
Even though the Location B is hidden, the query will find the Content because both conditions are satisfied:
```

- the Content item has Location B
- the Content item is visible (it has the visible Location A)

> even though Location B is hidden.

Reviewer 2: ```suggestion

```

## Sorting results

To sort the results of a query, use one of more [Sort Clauses](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.2/eZ/Publish/API/Repository/Values/Content/Query/SortClause.php).

For example, to order search results by their publicationg date, from oldest to newest,
and then alphabetically by content name, add the following Sort Clauses to the query:

``` php
$query->sortClauses = [
    new SortClause\DatePublished(LocationQuery::SORT_ASC),
    new SortClause\ContentName(LocationQuery::SORT_DESC),
];
```

!!! tip

>     See [Sort Clause reference](../guide/search.md#sort-clauses-reference) for full list and details of available Sort Clauses.

Reviewer 2: ```suggestion
    For the full list and details of available Sort Clauses, see [Sort Clause reference](../guide/search.md#sort-clauses-reference).
```
